### PR TITLE
Card Management System

### LINKED ISSUE #56: As a creator I can create a lesson
So I can add cards to the lesson and use it to test myself in quiz mode

### LINKED ISSUE #57: As a creator I can delete a lesson
So I can discard a lesson I am no longer using

### LINKED ISSUE #95: As a creator I can view all lessons
So I can at a glance find out what lessons are there, and proceed to add, delete or rename them.

### PR DESCRIPTION
**Pull request by DigiPie**
### Implement Card Management System (CMS)
This is a follow up to the PR: **Enhance Lesson class and refactor Card class #77** which was merged earlier in mid-v1.2 implementation.

#### Add CMS-related commands
1. Implements `addLesson` command for users to add lessons (closes #56)
2. Implements `listLessons` command for users to list all lessons (closes #95)
 3. Implements `deleteLesson` command for users to delete lessons (closes #57)

#### Major (but incremental) overhaul of `package seedu.address.logic` design and implementation
1. Rename `Model` interface to `ManagementModel` and  `ModelManager` class to `ManagementModelManager` given that there are equivalent `QuizModel` interface and `QuizModelManager` class
2. **Add `Model` interface which `ManagementModel` and `QuizModel` implements and refactor `ManagementModel` and `QuizModel` related files**
3. **Add `Command` interface and refactor all commands to implement it**
4. Delete `Command` and `QuizCommand` abstract classes given that all commands implement `Command` interface instead
5. Update all commands which use the `Model` passed into the `execute()` override method to check if the models passed in are `ManagementModel` or `QuizModel` objects
6. Rename `BrainTrainParser` class to `ManagementModeParser` given that there is an equivalent `QuizModeParser` class
7. **Add `Parser` interface which `ManagementModeParser` and `QuizModeParser` implements and refactor `ManagementModeParser` and `QuizModeParser` related files**
8. Remove `MissingCoreException` and `MissingOptionalException` as they are no longer in use.

#### Documents
1. Update `UserGuide.adoc`'s Introduction and Quick Start sections
2. Update `digipie.adoc` with v1.2 contributions.
3. Update terms in `DeveloperGuide.adoc` and `LearningOutcomes.adoc` to reflect refactoring of `package seedu.address.logic` files

### CODE REVIEW

File: docs/UserGuide.adoc [hunk 1]
```diff
@@ -77,6 +77,24 @@ Examples:
...
+Format: `addLesson n/NAME c/CORE... [o/OPTIONAL]...`
+
+Usage rules:
+
+* The lesson to be added must contain at least 2 `CORE` parameters. These parameters are `CORE` headers which define the lesson cards' `CORE` values. An example of a `CORE` header and a card's `CORE` value would be 'Country' and 'Japan' respectively.
+
+[TIP]
+The lesson can have any number of `OPTIONAL` headers (including 0).
+
+The `OPTIONAL` headers are optional. When added, these headers are displayed during quiz mode when <<hint, `\hint`>> is entered.
```
Comment (halcyoneee) [inline docs/UserGuide.adoc]: the values of these `OPTIONAL` headers are displayed

File: src/main/java/seedu/address/logic/parser/Syntax.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package seedu.address.logic.parser;
+
+/**
+ * Contains Command Line Interface (CLI) syntax definitions common to multiple commands
+ */
+public class Syntax {
```
Comment (halcyoneee) [inline src/main/java/seedu/address/logic/parser/Syntax.java]: Keep track of all the rename you did from AB4 files in the PR, as others such as Lei maybe using it as well.

Comment (DigiPie) [reply src/main/java/seedu/address/logic/parser/Syntax.java]: Alright! @lallanachang do take note that there is this `Syntax.java` where we are supposed to put our commands' argument prefixes in.

File: docs/UserGuide.adoc (hunk 1 above)
Comment (DigiPie) [reply docs/UserGuide.adoc]: Thanks for spotting this @halcyoneee :) I have rectified the error.

File: src/main/java/seedu/address/logic/commands/exceptions/CommandException.java [hunk 3]
```diff
@@ -1,9 +1,14 @@
 package seedu.address.logic.commands.exceptions;
 
 /**
- * Represents an error which occurs during execution of a {@link Command}.
+ * Represents an error which occurs during execution of a {@link seedu.address.logic.commands.Command}.
```
Comment (halcyoneee) [inline src/main/java/seedu/address/logic/commands/exceptions/CommandException.java]: Change the full path package if not necessary

File: src/main/java/seedu/address/logic/commands/management/DeleteLessonCommand.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Deletes the lesson identified by its index number in the numbered list"
+            + " shown when the command \'listLessons\' is entered.\n"
+            + "Parameter: INDEX (must be a positive integer)\n"
+            + "Example: " + COMMAND_WORD + " 1";
+    /**
+     * Feedback message displayed to the user upon successful execution of this command
+     */
+    public static final String MESSAGE_SUCCESS = "Deleted lesson: ";
```
Comment (halcyoneee) [inline src/main/java/seedu/address/logic/commands/management/DeleteLessonCommand.java]: To standardise with all the rest of the classes, include `%1$s`, such that it is `Deleted lesson: %1$s\n`

File: src/main/java/seedu/address/logic/commands/exceptions/CommandException.java (hunk 3 above)
Comment (halcyoneee) [inline src/main/java/seedu/address/logic/commands/exceptions/CommandException.java]: Similar to previous comment, change it back to Command

File: src/test/java/seedu/address/logic/commands/ListLessonsCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+         * @param answer user input
+         */
+        public void updateTotalAttemptsAndStreak(int index, String answer) {
+            throw new AssertionError("This method should not be called.");
+        }
+
+        /**
+         * Returns if User is done with {@code Quiz}.
+         */
+        public boolean isDone() {
```
Comment (halcyoneee) [inline src/test/java/seedu/address/logic/commands/ListLessonsCommandTest.java]: Take note that this has been refactored to `isQuizDone` and `updateTotalAttemptsAndStreak` will return boolean instead in #96

**Comment by DigiPie** [comment]
> 
> 
> Great job on refactoring and keeping a high coverage on your parts. Also, great use of stub in your tests. As those comments as minor changes, do take note of the changes that need to be made in the upcoming milestone.

Thanks for your review @halcyoneee . I have made the required changes and will push them in a subsequent commit.

**Pull request merged**